### ISSUE TITLE
TODO: Fix InvokeFunction to deal with parameters that are collections of Enums

### ISSUE THREAD

**Issue by dnicolauv** [JimDaly/CRMJSWebAPIServiceUtil#25]
When trying to use a function with a parameter of type enum I'm really having a hard time trying to use it. Looking at the code generated it doesn't seem as the invokefunction method is ready to handle an incoming parameter of enum type. Actually "isEnum" is always undefined. But even if it wasn't it doesn't look like it would handle it correctly anyway. Have you tested it? Am i missing something? Can you provide an example?

Thank you

**Comment by JimDaly** [comment]
Which function are you using?
Which version of the Web API?

**Comment by dnicolauv** [comment]
Im using latest version, just downloaded it some days ago. Im trying the QuerySchedule function, which has a param of type TimeCode enum array...

**Comment by JimDaly** [comment]
Not asking about the version of this utility - looking for the version of Dynamics 365.
Latest version of Web API is v8.2.
You can confirm by querying <server uri>/api/data/v8.0/RetrieveVersion in your browser.

**Comment by dnicolauv** [comment]
Its a dyn365 online trial, so 8.2

**Comment by JimDaly** [comment]
This is what the generated QuerySchedule function looks like where my namespace is Sdk.Test:
```
  /**
    * @function Sdk.Test.QuerySchedule
    * @memberOf Sdk.Test
    * @description Searches the specified resource for an available time block that matches the specified parameters.
    * @param {Guid} resourceId The resource.
    * @param {Date} start The start time for the scheduled block of time.
    * @param {Date} end The end time for the scheduled block of time.
    * @param {Array} timeCodes The time codes to look for: Available, Busy, Unavailable, or Filter, which correspond to the resource IDs
    * @param {String} [callerId] A string representation of the GUID value for the user to impersonate.
    * @returns {Promise} A Sdk.Test.QueryScheduleResponse instance when resolved or an Error if rejected. 
    */
    this.QuerySchedule = function (
        resourceId,
        start,
        end,
        timeCodes,
        callerId
        ) {
        if(!isGuid(resourceId)) {
            throw new Error(NS + ".QuerySchedule resourceId parameter must be a Guid value.");
        }
        if(!isDate(start)) {
            throw new Error(NS + ".QuerySchedule start parameter must be a Date value.");
        }
        if(!isDate(end)) {
            throw new Error(NS + ".QuerySchedule end parameter must be a Date value.");
        }
        if(!isArrayOfEnumMember(Sdk.Test.TimeCode,timeCodes)) {
            throw new Error(NS + ".QuerySchedule timeCodes parameter must be a Array value.");
        }
        if (!isOptionalGuid(callerId)) {
            throw new Error(NS+".QuerySchedule callerId parameter must be a string representation of a GUID value, null or undefined.");
        }
        var parameters = [];
        if (resourceId) {
        parameters.push({ "ResourceId": resourceId });
        }
        if (start) {
        parameters.push({ "Start": start, isDate: true  });
        }
        if (end) {
        parameters.push({ "End": end, isDate: true  });
        }
        if (timeCodes) {
        parameters.push({ "TimeCodes": timeCodes });
        }
        return invokeFunction("QuerySchedule",parameters,null,null,null,callerId,Sdk.Test.QueryScheduleResponse);
    }
```
In this case, the **TimeCodes** parameter is an array of enum values, not just an enum value. The isEnum flag is passed when the library needs to explicitly flag a parameter as an enum and to differentiate them from strings since enum values are passed as strings rather than numbers.

Can you share the code you are using to call QuerySchedule?
What kind of error are you getting?

**Comment by dnicolauv** [comment]
This is the code I'm using. You can see some syntax I've tried and what each returns:
`var RAOApi = demoWebApi.RAO;

    var timecodes = [];

    //timecodes.push({ "TimeCode": "Microsoft.Dynamics.CRM.TimeCode'" + "Available" + "'", isEnum: true }); // returns invalid JSON
    //timecodes.push({ "TimeCode": "Microsoft.Dynamics.CRM.TimeCode'" + RAOApi.Available + "'", isEnum: true }); // returns invalid JSON
    //timecodes.push(demoWebApi.RAO.TimeCode.Available); // returns 500 Internal Server Error : "Object reference not set to an instance of an object."
    timecodes.push("Unavailable"); // returns 500 Internal Server Error : "Object reference not set to an instance of an object."

    RAOApi.QuerySchedule("4bce2a2b-1b2f-486d-90fc-f628c3e19a07", new Date(), new Date(), timecodes, "4bce2a2b-1b2f-486d-90fc-f628c3e19a07").then(function (resp)
    {
        console.log(resp);
        console.log(resp.TimeInfos.length);
    });`

By the way I've tested this call with this params in a console .net application using the crm service client and they return data, so I discard a data issue. This returns correct values:
`QueryScheduleRequest scheduleRequest = new QueryScheduleRequest
            {
                ResourceId =new Guid("4bce2a2b-1b2f-486d-90fc-f628c3e19a07"),
                Start = DateTime.Now,
                End = DateTime.Today.AddDays(7),
                TimeCodes = new TimeCode[] { TimeCode.Available }
            };
            QueryScheduleResponse scheduleResponse = (QueryScheduleResponse)organizationService.Execute(scheduleRequest);`

**Comment by JimDaly** [comment]
OK. I think I see the issue.

Thanks for reporting this. As I mentioned in the notes, I haven't tested every operation and it looks like this is a missed combination of parameters.

Ultimately, the responsibility of the generated **QuerySchedule** and **invokeFunction** functions is to compose a URL for the function which I expect should look _something like_ this:

`[Server URI]/api/data/v8.2/QuerySchedule(ResourceId=@p1,Start=@p2,End=@p3,TimeCodes=@p4)?@p1='4bce2a2b-1b2f-486d-90fc-f628c3e19a07'&p2=2017-05-06T22:24:55Z&&p3=2017-05-07T22:24:55Z&p4=[Microsoft.Dynamics.CRM.TimeCode.Available]`

 > In your example you are also passing in the callerID, which happens to be the same value as the resource, but that seems to be fine.

As you have suggested, it looks like it needs to massage the enum values to append the namespace since these values are being passed through the URL.

I think **QuerySchedule** needs to flag this parameter as an array of enum values, and **invokeFunction** needs to look for this flag so that it can loop through the values in the array and add the appropriate namespace.

**Comment by dnicolauv** [comment]
Yes, looks like so. For now I think I can solve my specific need with the SearchRequest function instead (which works great and was straightforward to use in your lib), but it would be great to have this one solved as well. I think the most important thing would be to be sure about how you are supposed to exactly append that to the url, being an array of enums...

**Comment by JimDaly** [comment]
Agreed. Since this is a GET, it should be pretty straightforward to test since you can just send it with your browser after you have logged in. The URL should be some variation of this:
`[Server URI]/api/data/v8.2/QuerySchedule(ResourceId=@p1,Start=@p2,End=@p3,TimeCodes=@p4)?@p1='4bce2a2b-1b2f-486d-90fc-f628c3e19a07'&p2=2017-05-06T22:24:55Z&&p3=2017-05-07T22:24:55Z&p4=[Microsoft.Dynamics.CRM.TimeCode.Available]`
The tricky parts are:
1. Formatting of the datetimeoffset values. The example above doesn't include hours time zone. Probably this: [Passing DateTimeOffset as WebAPI query string](https://stackoverflow.com/questions/28549737/passing-datetimeoffset-as-webapi-query-string)
2. How to pass the enum values? With or without the fully qualified name? In quotes or not? I can say I'm quite sure the number values associated with the enum are not used. Could be Microsoft.Dynamics.CRM.TimeCode'Available'?
Wish I could tell you off the top of my head exactly what to use. But I would attempt this as a trial and error to figure this out.  I will ask some people so that we can cover this in [Use Web API functions > Passing parameters to a function](https://msdn.microsoft.com/en-us/library/gg309638.aspx#bkmk_passParametersToFunctions)

**Comment by abhisheksharma0994** [comment]
Hi,

Even I am stuck in the Web api Enum part. Any update on this will be really helpful :)

**Comment by AshleyAsh90** [comment]
Hello, I was wondering if there are any updates on this?

**Comment by JimDaly** [comment]
I haven't been able to find the time to dig into this. The issue appears to be how to pass through a collection of enum values. Fixing this would be easier with some help. I don't happen to have a project that requires me to use this function. 

If you are blocked by this, I expect you have an environment with data set up for testing that I don't.

You should be able to modify the generated library function to try a few options.

If you are passing a single value to the **QuerySchedule** function, the **TimeCodes** parameter is usually set to look for schedules that match the `Available` enum.

Remove the validation checking code and pass through the raw value. 

The validation checking code for the **TimeCodes** parameter within the generated **QuerySchedule** function should look something like this:

```
        if(!isArrayOfEnumMember(Sdk.Test.TimeCode,timeCodes)) {
            throw new Error(NS + ".QuerySchedule timeCodes parameter must be a Array value.");
        }
```

The raw value could be one of these:

`[Microsoft.Dynamics.CRM.TimeCode.Available]`
`["Microsoft.Dynamics.CRM.TimeCode.Available"]`
`["Microsoft.Dynamics.CRM.TimeCode.'Available'"]`


If you can verify which style should be used for the values, that will help me correct the code that is included in the generated library function.

**Comment by AshleyAsh90** [comment]
Actually I am using WebAPI calls using RestSharp. Which returns an error 
"The child type 'Microsoft.Dynamics.CRM.TimeCode.Available' in a cast was not an entity type. Casts can only be performed on entity types." It shows as an Odata Exception

**Comment by JimDaly** [comment]
@AshleyAsh90  - This project is about a JavaScript/TypeScript code generation utility. It is a personal side project of mine from a couple years back. If you have general questions about how to work with Web API, I recommend that you submit it to the relevant content topics on docs.microsoft.com.

For example, a relevant topic under
[Use the Common Data Service for Apps Web API](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/overview) or [Web API types and operations](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/web-api-types-operations) or [Use Web API functions](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/use-web-api-functions)

**Comment by AshleyAsh90** [comment]
I apologize. I didn't notice that.
